deprecate low level API

alexlamsl opened the pull request:
fixes #1872

After this PR, the following API remains:
- `AST_*`
- `describe_ast()`
- `Dictionary` (name cache manipulation?)
- `minify()`
- `_push_uniq()` (only `bin/uglifyjs` needs this)
- `TreeWalker`

alexlamsl commented:
@kzc so I've basically made a copy of `tools/node.js` and `tools/exports.js` under `test/` which exports more internal functions, so as to keep changes to that folder to a minimum.

As for the public APIs, I've only left `parse()` and `Compressor()` exposed but removed the rest. This isn't final, as I'm sure you have a better idea than me on what to keep/remove :wink:

kzc commented:
That should do the trick. 

It'd better not to export the low level API at all but we want the internal tests to continue to work. 

Since the low level API is deprecated with this PR, any thought given to having minify() optionally accept uglify AST and optionally produce uglify AST?  If the first argument to minify() is `instanceof AST_Node` that would handle the AST input. If the option `output.ast` is true-ish then `result.ast` could be set to the output uglify AST.

alexlamsl commented:
> It'd better not to export the low level API at all but we want the internal tests to continue to work.

We don't actually need to expose `Compressor`, as we can then remove the [test for legacy API](https://github.com/mishoo/UglifyJS2/blob/ec443e422c220619fe671166e467d3d8838b0cab/test/mocha/minify.js#L187-L194) since we no longer export anything enough to be abused.

(`parse()` is still needed since `bin/uglifyjs` need to parse its `--compress keep_fargs=0`)

> Since the low level API is deprecated with this PR, any thought given to having minify() optionally accept uglify AST and optionally produce uglify AST?

I am open to suggestions - do people do this? The only worry I have is with programmatically constructed AST not being compatible with those generated from `parse()`, since no(t much) validation takes place with `AST_Node` constructors.

kzc commented:
@alexlamsl Regarding the minify/AST thing, if the low level API goes away, then how do users get to the AST to do custom searches, etc?  You mentioned you have projects that traverse the AST. 

I would not worry about AST validation - if they tread down this path they ought to be aware of the pitfalls. It would mostly be for reading the AST, or inspecting the AST between minify() invocations with different options.

alexlamsl commented:
> Regarding the minify/AST thing, if the low level API goes away, then how do users get to the AST to do custom searches, etc? You mentioned you have projects that traverse the AST.

`TreeWalker` is exposed alongside `AST_*`, which is what I currently use in my projects to pre-process stuff before throwing it into `minify()`. But yes it's a bit inefficient, as I rely on `parse()` and then `print_to_string()`.

You did raise a good point - if I can output AST from `minify()`, I can remove exposing `parse()` since it's just `minify(code, {compress:false,mangle:false,output:AST_Node})`.

kzc commented:
> minify(code, {compress:false,mangle:false,output:AST_Node})

actually:
```
var result = minify(code, { compress:false, mangle:false, output:{ast:true} })
```
... inspect/traverse `result.ast` ...
```
var result = minify(result.AST, { compress:true, mangle:true, output:{ast:true} })
```

alexlamsl commented:
Turns out we already support AST as input to `minify()`:
```js
$ node
> var U = require('uglify-js');
> U.minify(U.parse("console.log(true);")).code;
'console.log(!0);'
```

So it's only `output:{ast: true}` that we need to worry about. Now do we also output `result.code` or do we skip that stage altogether? The latter would mean reporting an error if `sourceMap` is specified, because there's no way you can get one without invoking `OutputStream`.

kzc commented:
> Turns out we already support AST as input to minify()

That wouldn't happen by accident - did you previously add support?

> So it's only output:{ast: true} that we need to worry about. Now do we also output result.code or do we skip that stage altogether? 

Good question. How about adding another output option defaulting to true: `output.code` to emit the code string?  That way one could emit only the AST with `output: { ast: true, code: false }`.

kzc commented:
A syntax verifier could use the following minify() option: 
`{ warnings: true, mangle: false, compress: false, output: { ast: false, code: false } }`
as it would not need the output code, nor the AST.

alexlamsl commented:
> That wouldn't happen by accident - did you previously add support?

I have :tropical_fish: brain.

kzc commented:
This link does a great job of explaining how javascript strings are represented *within* most JS engines:

https://mathiasbynens.be/notes/javascript-encoding

It's bizarre - some variant of UCS-2.

alexlamsl commented:
> It's bizarre - some variant of UCS-2.

Yup - and then Java came along and taught me NFC, NFD, NFKC & NFKD.

I almost hung myself back then.

alexlamsl commented:
Rebased and updated to remove `Compressor` and `parse()` from public API.

After this, I'll spin up a replacement for #769. (Then if time permits, have a dig through `harmony` for fixing #1185...)

kzc commented on tools/exports.js:
```diff
@@ -1,15 +1,4 @@
-exports["Compressor"] = Compressor;
 exports["Dictionary"] = Dictionary;
-exports["JS_Parse_Error"] = JS_Parse_Error;
```
I think some uglify 2.x wrappers need to access `JS_Parse_Error` for some reason. With the new 3.x minify API perhaps this is no longer an issue.

kzc replied:
Can `minify()` still throw uncaught exceptions?

alexlamsl replied:
`minify()` can still throw `JS_Parse_Error`, but the way to test for it, at least in `bin/uglifyjs`, is to check for `.name` instead of `instanceof`

https://github.com/mishoo/UglifyJS2/pull/1877/files#diff-e5511a595da8371c33a33346f3e22fd4L190

kzc replied:
```patch
-        if (ex instanceof UglifyJS.JS_Parse_Error) {
+        if (ex.name == "SyntaxError") {
```
Ah - I see. Is `exception.name` a standard ECMAScript thing?

kzc replied:
Would I be a heretic to suggest a new `minify()` option `{ catch: true }` and emit the exception to `result.errors` or `result.exception` or equivalent?

alexlamsl replied:
So instead of throwing, return with an exception instead?

I suppose it's not a bad idea, seeing that people don't like to wrap `try-catch` all the time if they can avoid it. The quick patch I had in mind can't return anything else than the `Error` instance though, so no point wrapping it within another object as the return value?

alexlamsl replied:
> Ah - I see. Is  exception.name  a standard ECMAScript thing?

All the standard `Error` sub-classes has their own `.name`, and the field is also mentioned by MDN:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/name

alexlamsl commented:
I shall address https://github.com/mishoo/UglifyJS2/pull/1877#discussion_r115150972 in a new PR.
